src/snarks/groth16/mpc/phase2.hpp

```diff
@@ -0,0 +1,309 @@
+
+    srs_mpc_phase2_challenge(
+        const srs_mpc_hash_t transcript_digest,
+        srs_mpc_phase2_accumulator<ppT> &&accumulator);
+
+    bool operator==(const srs_mpc_phase2_challenge<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    static srs_mpc_phase2_challenge<ppT> read(std::istream &in);
+};
+
+/// Reponse produced by participant in Phase2 of the SRS generation MPC.
+template<typename ppT> class srs_mpc_phase2_response
+{
+public:
+    srs_mpc_phase2_accumulator<ppT> new_accumulator;
+    srs_mpc_phase2_publickey<ppT> publickey;
+
+    srs_mpc_phase2_response(
+        srs_mpc_phase2_accumulator<ppT> &&new_accumulator,
+        srs_mpc_phase2_publickey<ppT> &&publickey);
+
+    bool operator==(const srs_mpc_phase2_response<ppT> &other) const;
+    bool is_well_formed() const;
+    void write(std::ostream &out) const;
+    static srs_mpc_phase2_response<ppT> read(std::istream &in);
+};
+
+// Phase2 functions
+
+template<mp_size_t n, const libff::bigint<n> &modulus>
+void srs_mpc_digest_to_fp(
+    const srs_mpc_hash_t transcript_digest,
+    libff::Fp_model<n, modulus> &out_fr);
+
+template<typename ppT>
+libff::G2<ppT> srs_mpc_digest_to_g2(const srs_mpc_hash_t digest);
+
+/// Given the output from the linear combination of the L1 layer of the SRS
+/// circuit, compute the starting parameters for Phase 2 (the MPC for C2
+/// layer).  See "Initialization" in section 7.3 of [BoweGM17].
+template<typename ppT>
+srs_mpc_phase2_accumulator<ppT> srs_mpc_phase2_begin(
+    const srs_mpc_hash_t cs_hash,
+    const srs_mpc_layer_L1<ppT> &layer_L1,
+    size_t num_inputs);
+
+/// Outputs the public key (which includes the POK) for our secret.  Correponds
+/// to steps 1 and 2 in "Computation", section 7.3 of [BoweGM17]
+template<typename ppT>
+srs_mpc_phase2_publickey<ppT> srs_mpc_phase2_compute_public_key(
+    const srs_mpc_hash_t transcript_digest,
+    const libff::G1<ppT> &last_delta,
+    const libff::Fr<ppT> &secret);
+
+/// Verifies that a public key is correct, given some previous delta.
+/// Corresponds to steps 1 and 2 from "Verfication" in section 7.3 of
+/// [BoweGM17] (for a single contributor $j$).  Note that the caller is
+/// responsible for checking the transcript_hash.
+template<typename ppT>
+bool srs_mpc_phase2_verify_publickey(
+    const libff::G1<ppT> last_delta_g1,
+    const srs_mpc_phase2_publickey<ppT> &publickey);
+
+/// Core update function, which applies a secret contribution to an
+/// accumulator.  Corresponds to steps 3 onwards in "Computation",
+/// section 7.3 of [BoweGM17].
+template<typename ppT>
+srs_mpc_phase2_accumulator<ppT> srs_mpc_phase2_update_accumulator(
+    const srs_mpc_phase2_accumulator<ppT> &last_accum,
+    const libff::Fr<ppT> &delta_j);
+
+/// Assuming last is fully verified, and updated.delta_g1 has the appropriate
+/// ratio, check that all other elements of updated.  This covers the G2 part
```

Comment:
`check that all other elements of updated.` -> Unfinished sentence?

Reply:
Fixed sentence